core/src/test/java/org/apache/iceberg/io/TestAppenderFactory.java

```diff
@@ -0,0 +1,265 @@
+  }
+
+  protected abstract FileAppenderFactory<T> createAppenderFactory(List<Integer> equalityFieldIds,
+                                                                  Schema eqDeleteSchema,
+                                                                  Schema posDeleteRowSchema);
+
+  protected abstract T createRow(Integer id, String data);
+
+  protected abstract StructLikeSet expectedRowSet(Iterable<T> records) throws IOException;
+
+  protected abstract StructLikeSet actualRowSet(String... columns) throws IOException;
+
+  private OutputFileFactory createFileFactory() {
+    return new OutputFileFactory(table.spec(), format, table.locationProvider(), table.io(),
+        table.encryption(), 1, 1);
+  }
+
+  private PartitionKey createPartitionKey() {
+    if (table.spec().isUnpartitioned()) {
+      return null;
+    }
+
+    Record record = GenericRecord.create(table.spec().schema()).copy(ImmutableMap.of("data", "aaa"));
+
+    PartitionKey partitionKey = new PartitionKey(table.spec(), table.schema());
+    partitionKey.partition(record);
+
+    return partitionKey;
+  }
+
+  private List<T> testRowSet() {
+    return Lists.newArrayList(
+        createRow(1, "aaa"),
+        createRow(2, "bbb"),
+        createRow(3, "ccc"),
+        createRow(4, "ddd"),
+        createRow(5, "eee")
+    );
+  }
+
+  private DataFile prepareDataFile(List<T> rowSet, FileAppenderFactory<T> appenderFactory,
+                                   OutputFileFactory outputFileFactory) throws IOException {
+    DataWriter<T> writer = appenderFactory.newDataWriter(outputFileFactory.newOutputFile(), format, partition);
+    try (DataWriter<T> closeableWriter = writer) {
+      for (T row : rowSet) {
+        closeableWriter.add(row);
+      }
+    }
+
+    return writer.toDataFile();
+  }
+
+  @Test
+  public void testDataWriter() throws IOException {
+    FileAppenderFactory<T> appenderFactory = createAppenderFactory(null, null, null);
+    OutputFileFactory outputFileFactory = createFileFactory();
+
+    List<T> rowSet = testRowSet();
+    DataFile dataFile = prepareDataFile(rowSet, appenderFactory, outputFileFactory);
+
+    table.newRowDelta()
+        .addRows(dataFile)
+        .commit();
+
+    Assert.assertEquals("Should have the expected records.", expectedRowSet(rowSet), actualRowSet("*"));
+  }
+
+  @Test
+  public void testEqDeleteWriter() throws IOException {
+    List<Integer> equalityFieldIds = Lists.newArrayList(table.schema().findField("id").fieldId());
+    FileAppenderFactory<T> appenderFactory = createAppenderFactory(equalityFieldIds,
+        table.schema().select("id"), null);
+    OutputFileFactory outputFileFactory = createFileFactory();
+
+    List<T> rowSet = testRowSet();
+    DataFile dataFile = prepareDataFile(rowSet, appenderFactory, outputFileFactory);
+
+    table.newRowDelta()
+        .addRows(dataFile)
+        .commit();
+
+    List<T> deletes = Lists.newArrayList(
+        createRow(1, "aaa"),
+        createRow(3, "bbb"),
```

Comment:
If the data here is `bbb` instead of `ccc` on purpose, then could you add a comment that this is testing that just `id` is used for comparison?